File: nl.naturalis.nba.client.php/model/TaxonomicStatus.php
<?php

namespace nl\naturalis\nba\client\php\model;

final class TaxonomicStatus {

	private static $acceptedName;
	private static $synonym;
	private static $basionym;
	private static $homonym;
	private static $ambiguousSynonym;
	private static $misappliedName;
	private static $misspelledName;
	private static $provisionallyAccepted;

	public static function ACCEPTED_NAME()
	{
		if (self::$acceptedName === null)
			self::$acceptedName = new TaxonomicStatus("accepted name");
		return self::$acceptedName;
	}

	public static function SYNONYM()
	{
		if (self::$synonym === null)
			self::$synonym = new TaxonomicStatus("synonym");
		return self::$synonym;
	}

	public static function BASIONYM()
	{
		if (self::$basionym === null)
			self::$basionym = new TaxonomicStatus("basionym");
		return self::$basionym;
	}

	public static function HOMONYM()
	{
		if (self::$homonym === null)
			self::$homonym = new TaxonomicStatus("homonym");
		return self::$homonym;
	}

	public static function AMBIGUOUS_SYNONYM()
	{
		if (self::$ambiguousSynonym === null) {
			self::$ambiguousSynonym = new TaxonomicStatus("ambiguous synonym");
		}
		return new TaxonomicStatus("ambiguous synonym");
	}

	public static function MISAPPLIED_NAME()
	{
		if (self::$misappliedName === null)
			self::$misappliedName = new TaxonomicStatus("misapplied name");
		return self::$misappliedName;
	}

	public static function MISSPELLED_NAME()
	{
		if (self::$misspelledName === null)
			self::$misspelledName = new TaxonomicStatus("misapplied name");
		return self::$misspelledName;
	}

	public static function PROVISIONALLY_ACCEPTED()
	{
		if (self::$provisionallyAccepted === null)
			self::$provisionallyAccepted = new TaxonomicStatus("provisionally accepted name");
		return self::$provisionallyAccepted;
	}

	private final $name;

	private function __construct(string $name)
	{
		$this->name = $name;
	}

	public function __toString()
	{
		return name;
	}
}

File: nl.naturalis.nba.api/src/main/java/nl/naturalis/nba/api/model/TaxonomicStatus.java
package nl.naturalis.nba.api.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonValue;

public enum TaxonomicStatus implements INbaModelObject
{
	ACCEPTED_NAME("accepted name"),
	ALTERNATIVE_NAME("alternative name"),
	AMBIGUOUS_SYNONYM("ambiguous synonym"),
	BASIONYM("basionym"),
	HOMONYM("homonym"),
	INVALID_NAME("invalid name"),
	MISAPPLIED_NAME("misapplied name"),
	MISIDENTIFICATION("misidentification"),
	MISSPELLED_NAME("misspelled name"),
	NOMEN_NUDUM("nomen nudum"),
	PREFERRED_NAME("preferred name"),
	PROVISIONALLY_ACCEPTED("provisionally accepted name"),
  SYNONYM("synonym");

	@JsonCreator
	public static TaxonomicStatus parse(@JsonProperty("name") String name)
	{
		if (name == null) {
			return null;
		}
		for (TaxonomicStatus status : TaxonomicStatus.values()) {
			if (status.name.equalsIgnoreCase(name)) {
				return status;
			}
		}
		throw new IllegalArgumentException("Invalid taxonomic status: " + name);
	}

	private final String name;

	private TaxonomicStatus(String name)
	{
		this.name = name;
	}

	private TaxonomicStatus()
	{
		this.name = name().toLowerCase();
	}

	@JsonValue
	public String toString()
	{
		return name;
	}

}